### PR TITLE
fix: correctly detect changed data

### LINKED ISSUE #8157: `$page.data` does not get updated correctly
### Describe the bug

If the `data` returned from `load` inside a `+page.server.js` ("page-X") has the same shape as the data returned inside its corresponding `+layout.server.js`, the `$page.data` does not get updated when routing from "page-X" to another page that does not have its own `+page.server.js` file.

As a workaround I'm currently just adding a random key to the return value of `+layout.server.js`.

### Reproduction

https://stackblitz.com/edit/sveltejs-kit-template-default-bcif9n?file=src/routes/page-2/+page.server.js

1. toggle between the `root` and `page-1` links a few time
2. the heading at the top should update accordingly
3. now toggle between `root` and the `page-2` links
4. the heading does not get updated when clicking on `root`
5. now add `other: 5` to the return object in `/routes/+page.layout.js`
6. now toggling between `root` and `page-2` works
7. but toggling between `root` and `page-1` does not update the heading

### Logs

_No response_

### System Info

```Shell
Stackblitz
```

### Severity

serious, but I can work around it

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by Rich-Harris**
fixes #8157. this really needs some tests but my brain is melted right now

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

File: packages/kit/src/runtime/client/client.js [hunk 1]
```diff
@@ -446,16 +445,26 @@ export function create_client({ target, base }) {
...
-			if (data_changed || !current.branch.some((previous) => previous === node)) {
-				result.props[`data_${i}`] = data;
-				data_changed = data_changed || Object.keys(node.data ?? {}).length > 0;
+			if (data_changed) {
+				result.props[`data_${p}`] = data;
 			}
+
+			p += 1;
 		}
+
```
[the remark this answers is not in the document]
Comment (Rich-Harris) [reply packages/kit/src/runtime/client/client.js:467]: ah yep, you're right. removed

**Comment by dummdidumm** [comment]
You said you wanted to add some tests - still up for it? I could take a look if you want, too.

File: .changeset/grumpy-adults-breathe.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@sveltejs/kit': patch
+---
+
+[fix] correctly detect changed data
```
Comment (benmccann) [inline .changeset/grumpy-adults-breathe.md:5]:
```suggestion
fix: correctly detect changed data
```

**Comment by Rich-Harris** [comment]
@dummdidumm yeah, I was planning to (but don't let me stop you from getting there first if you're feeling inspired 😛 )

**Comment by dummdidumm** [comment]
While writing a test I discovered that this doesn't fix the issue yet; we have to do a `Math.max(branchLengthOld, branchLengthNew)`, else a now shorter branch length with the old one containing a data point at a higher index would give a false negative (didn't change when it actually did)

**Comment by dummdidumm** [comment]
While writing a test I discovered that this doesn't fix the issue yet; we have to do a `Math.max(branchLengthOld, branchLengthNew)`, else a now shorter branch length with the old one containing a data point at a higher index would give a false negative (didn't change when it actually did). Pushing a fix.

**Review by dummdidumm — APPROVED** [review]

**Pull request merged**